distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/Manager.kt

```diff
@@ -0,0 +1,170 @@
+        leaderLatch.start()
+    }
+
+    private fun initCuratorCacheForManagerEvents(cache: CuratorCache, treeName: String) {
+        val cacheInitLocker = Semaphore(0)
+        cache.listenable().addListener(object : CuratorCacheListener {
+            override fun event(type: CuratorCacheListener.Type?, oldData: ChildData?, data: ChildData?) {
+                logger.trace { "nodeId=$nodeId $treeName event: type=$type, oldData=$oldData, data=$data" }
+                handleManagerEvent(ManagerEvent.ZK_WORKERS_CONFIG_CHANGED)
+            }
+
+            override fun initialized() {
+                cacheInitLocker.release()
+            }
+        })
+        cache.start()
+        cacheInitLocker.acquire()
+    }
+
+    private fun handleManagerEvent(newEvent: ManagerEvent) {
+        when (currentState.get()!!) {
+            State.IS_NOT_LEADER -> handleManagerEventAsNonLeader(newEvent)
+            State.IS_LEADER -> handleManagerEventAsLeader(newEvent)
+            State.SHUTDOWN -> {
+            }
+        }
+    }
+
+    private fun handleManagerEventAsLeader(event: ManagerEvent) {
+        when (event) {
+            ManagerEvent.ZK_WORKERS_CONFIG_CHANGED -> {
+                rebalanceAccumulator.publishEvent(RebalanceTrigger.DO_REBALANCE)
+            }
+            ManagerEvent.LEADERSHIP_LOST -> {
+                currentState.set(State.IS_NOT_LEADER)
+            }
+            ManagerEvent.SHUTDOWN -> {
+                currentState.set(State.SHUTDOWN)
+            }
+            ManagerEvent.LEADERSHIP_ACQUIRED -> {
+                logger.warn { "received ${ManagerEvent.LEADERSHIP_ACQUIRED} event, but manager is already the leader" }
+            }
+        }
+    }
+
+    private fun handleManagerEventAsNonLeader(event: ManagerEvent) {
+        when (event) {
+            ManagerEvent.ZK_WORKERS_CONFIG_CHANGED -> {
+            }
+            ManagerEvent.LEADERSHIP_ACQUIRED -> {
+                currentState.set(State.IS_LEADER)
+                rebalanceAccumulator.publishEvent(RebalanceTrigger.DO_REBALANCE)
+            }
+            ManagerEvent.SHUTDOWN -> {
+                currentState.set(State.SHUTDOWN)
+            }
+            ManagerEvent.LEADERSHIP_LOST -> {
+                logger.warn { "received ${ManagerEvent.LEADERSHIP_LOST} event, but manager has already lost leadership" }
+            }
+        }
+    }
+
+    override fun close() {
+        val managerStopTime = System.currentTimeMillis()
+        logger.info("Closing DJM manager entity...")
+
+        handleManagerEvent(ManagerEvent.SHUTDOWN)
+
+        aliveWorkersCache.close()
+        leaderLatch.close()
+        cleaner.close()
+
+        logger.info { "DJM manager was closed. Took ${System.currentTimeMillis() - managerStopTime} ms" }
+    }
+
+    private enum class ManagerEvent {
+        ZK_WORKERS_CONFIG_CHANGED, LEADERSHIP_LOST, LEADERSHIP_ACQUIRED, SHUTDOWN
+    }
+
+    private enum class RebalanceTrigger {
+        DO_REBALANCE
```

Comment:
I added this enum to emphasize that ManagerEvent is not "rebalance" event by itself:
`rebalanceAccumulator.publishEvent(ManagerEvent.ZK_WORKERS_CONFIG_CHANGED`